File: Medium_238. Product of Array Except Self.py
# From solutions

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        k=len(nums)
        j=[1]*k
        pre=1
        post=1
        for i in range(k):
            j[i]*=pre
            pre = pre*nums[i]
            j[k-i-1]*=post
            post=post*nums[k-1-i]
        return(j)

File: Medium_238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        std::vector<int> result_nums(nums.size(), 1);
        std::vector<int> left_mul(nums.size(), 1);
        std::vector<int> right_mul(nums.size(), 1);

        for(int i=nums.size()-2; i>=0; i--){
            right_mul[i] = right_mul[i+1]*nums[i+1];
        }
        
        for(int i=1; i<nums.size(); i++){
            left_mul[i] = left_mul[i-1]*nums[i-1];
            result_nums[i] = left_mul[i]*right_mul[i];
        }

        result_nums[0] = left_mul[0]*right_mul[0];

        return result_nums;

    }
};